src/proxy/http/balance.rs

```diff
@@ -118,19 +152,447 @@ where
+    where
+        Bal: svc::Service<T>,
+        Bal::Response: svc::Service<http::Request<A>> + HasEndpointStatus,
+        <<Bal as svc::Service<T>>::Response as svc::Service<http::Request<A>>>::Error: Into<Error>,
+        Bal::Error: Into<Error>,
+        R: rt::Make<router::Config> + Clone,
+    {
+        type Response = Service<R, Bal::Response, A>;
+        type Future = MakeFuture<R, Bal::Future, A>;
+        type Error = Bal::Error;
+
+        fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+            self.balance.poll_ready()
+        }
+
+        fn call(&mut self, target: T) -> Self::Future {
+            MakeFuture {
+                fallback: Some(self.fallback.clone()),
+                making: self.balance.call(target),
+                _marker: PhantomData,
+            }
+        }
+    }
+
+    impl<R, Bal, A> Clone for MakeSvc<R, Bal, A>
+    where
+        R: rt::Make<router::Config> + Clone,
+        Bal: Clone,
+    {
+        fn clone(&self) -> Self {
+            Self {
+                fallback: self.fallback.clone(),
+                balance: self.balance.clone(),
+                _marker: PhantomData,
+            }
+        }
+    }
+
+    impl<R, F, A> Future for MakeFuture<R, F, A>
+    where
+        F: Future,
+        F::Item: HasEndpointStatus,
+        R: rt::Make<router::Config>,
+    {
+        type Error = F::Error;
+        type Item = Service<R, F::Item, A>;
+        fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+            let balance = try_ready!(self.making.poll());
+            let status = balance.endpoint_status();
+            let fallback = self.fallback.take().expect("polled after ready");
+            Ok(Async::Ready(Service {
+                fallback,
+                status,
+                balance,
+                _marker: PhantomData,
+            }))
+        }
+    }
+
+    impl<R, Bal, A, B, C> svc::Service<http::Request<A>> for Service<R, Bal, A>
+    where
+        R: rt::Make<router::Config>,
+        R::Value: svc::Service<http::Request<A>, Response = http::Response<C>, Error = Bal::Error>,
+        Bal: svc::Service<http::Request<A>, Response = http::Response<B>>,
+        Bal::Error: Into<Error>,
+        B: Payload,
+        C: Payload<Error = B::Error>,
+    {
+        type Response = http::Response<Body<B, C>>;
+        type Error = Bal::Error;
+        type Future = future::Either<
+            future::Map<
+                <R::Value as svc::Service<http::Request<A>>>::Future,
+                fn(<R::Value as svc::Service<http::Request<A>>>::Response) -> Self::Response,
+            >,
+            future::Map<Bal::Future, fn(Bal::Response) -> Self::Response>,
+        >;
+
+        fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+            let ready = self.balance.poll_ready()?;
+            if !self.status.is_empty() {
+                // destroy the fallback router
+                self.fallback.destroy();
+            }
```

Comment:
don't we technically need to call poll_ready on the fallback here if status.is_empty?

Reply:
And, in fact, won't this prevent us from ever dispatching requests to the fallback while balance is not ready?

Reply:
@olix0r whoops, you're right --- fixed in 19cda744071c21b91decfe51e16c00d186df2559